File: Typescript/2024/2024-016/2024-016.ts
import * as fs from 'fs';
import * as path from 'path';
import PriorityQueue = require('js-priority-queue');

/**
 * Reads and parses the input file
 * @param filename - The path to the input file
 * @returns List of maze lines
 */
function parseInput(filename: string): string[] {
    const filePath = path.resolve(__dirname, filename);
    const data = fs.readFileSync(filePath, 'utf8');
    return data.trim().split('\n').map(line => line.trim());
}

/**
 * Solve Part 1: Find the lowest possible score
 * @param mazeLines - List of maze lines
 * @returns The lowest possible score
 */
function solveMaze(mazeLines: string[]): number {
    const directions = [
        [0, 1],   // East
        [1, 0],   // South
        [0, -1],  // West
        [-1, 0]   // North
    ];

    const rows = mazeLines.length;
    const cols = mazeLines[0].length;
    let start: [number, number] | null = null;
    let end: [number, number] | null = null;

    // Locate start and end positions
    for (let r = 0; r < rows; r++) {
        for (let c = 0; c < cols; c++) {
            if (mazeLines[r][c] === 'S') {
                start = [r, c];
            } else if (mazeLines[r][c] === 'E') {
                end = [r, c];
            }
        }
    }

    if (!start || !end) {
        throw new Error("Could not find 'S' or 'E' in the maze.");
    }

    const dist: number[][][] = Array.from({ length: rows }, () =>
        Array.from({ length: cols }, () => Array(4).fill(Infinity))
    );

    const pq = new PriorityQueue<[number, number, number, number]>({ comparator: (a, b) => a[0] - b[0] });
    dist[start[0]][start[1]][0] = 0;
    pq.queue([0, start[0], start[1], 0]); // cost, row, col, direction
    const visited = new Set<string>();

    while (pq.length > 0) {
        const [cost, r, c, d] = pq.dequeue();

        if (r === end[0] && c === end[1]) {
            return cost;
        }

        const stateKey = `${r},${c},${d}`;
        if (visited.has(stateKey)) continue;
        visited.add(stateKey);

        // Move forward
        const [dr, dc] = directions[d];
        const nr = r + dr;
        const nc = c + dc;
        if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && mazeLines[nr][nc] !== '#') {
            const newCost = cost + 1;
            if (newCost < dist[nr][nc][d]) {
                dist[nr][nc][d] = newCost;
                pq.queue([newCost, nr, nc, d]);
            }
        }

        // Turn left
        const leftDir = (d + 3) % 4;
        const leftCost = cost + 1000;
        if (leftCost < dist[r][c][leftDir]) {
            dist[r][c][leftDir] = leftCost;
            pq.queue([leftCost, r, c, leftDir]);
        }

        // Turn right
        const rightDir = (d + 1) % 4;
        const rightCost = cost + 1000;
        if (rightCost < dist[r][c][rightDir]) {
            dist[r][c][rightDir] = rightCost;
            pq.queue([rightCost, r, c, rightDir]);
        }
    }

    return -1; // No path found
}

/**
 * Solve Part 2: Count tiles on the best path
 * @param mazeLines - List of maze lines
 * @returns Number of tiles on the best path
 */
function solvePart2(mazeLines: string[]): number {
    const directions = [
        [0, 1],   // East
        [1, 0],   // South
        [0, -1],  // West
        [-1, 0]   // North
    ];

    const rows = mazeLines.length;
    const cols = mazeLines[0].length;
    let start: [number, number] | null = null;
    let end: [number, number] | null = null;

    // Locate start and end positions
    for (let r = 0; r < rows; r++) {
        for (let c = 0; c < cols; c++) {
            if (mazeLines[r][c] === 'S') {
                start = [r, c];
            } else if (mazeLines[r][c] === 'E') {
                end = [r, c];
            }
        }
    }

    if (!start || !end) {
        throw new Error("Could not find 'S' or 'E' in the maze.");
    }

    const dist: number[][][] = Array.from({ length: rows }, () =>
        Array.from({ length: cols }, () => Array(4).fill(Infinity))
    );

    const pq = new PriorityQueue<[number, number, number, number]>({ comparator: (a, b) => a[0] - b[0] });
    dist[start[0]][start[1]][0] = 0;
    pq.queue([0, start[0], start[1], 0]); // cost, row, col, direction
    const visited = new Set<string>();

    while (pq.length > 0) {
        const [cost, r, c, d] = pq.dequeue();

        const stateKey = `${r},${c},${d}`;
        if (visited.has(stateKey)) continue;
        visited.add(stateKey);

        // Move forward
        const [dr, dc] = directions[d];
        const nr = r + dr;
        const nc = c + dc;
        if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && mazeLines[nr][nc] !== '#') {
            const newCost = cost + 1;
            if (newCost < dist[nr][nc][d]) {
                dist[nr][nc][d] = newCost;
                pq.queue([newCost, nr, nc, d]);
            }
        }

        // Turn left
        const leftDir = (d + 3) % 4;
        const leftCost = cost + 1000;
        if (leftCost < dist[r][c][leftDir]) {
            dist[r][c][leftDir] = leftCost;
            pq.queue([leftCost, r, c, leftDir]);
        }

        // Turn right
        const rightDir = (d + 1) % 4;
        const rightCost = cost + 1000;
        if (rightCost < dist[r][c][rightDir]) {
            dist[r][c][rightDir] = rightCost;
            pq.queue([rightCost, r, c, rightDir]);
        }
    }

    // Trace back the best path
    const minCostEnd = Math.min(...dist[end[0]][end[1]]);
    if (minCostEnd === Infinity) {
        return 0;
    }

    const onBestPath: boolean[][] = Array.from({ length: rows }, () => Array(cols).fill(false));
    const queue: [number, number, number][] = [];
    for (let d = 0; d < 4; d++) {
        if (dist[end[0]][end[1]][d] === minCostEnd) {
            queue.push([end[0], end[1], d]);
        }
    }

    const visitedReverse = new Set<string>();

    while (queue.length > 0) {
        const [r, c, d] = queue.shift()!;
        onBestPath[r][c] = true;

        const costHere = dist[r][c][d];

        // Reverse move forward
        const nr = r - directions[d][0];
        const nc = c - directions[d][1];
        if (nr >= 0 && nr < rows && nc >= 0 && nc < cols) {
            if (mazeLines[nr][nc] !== '#' && dist[nr][nc][d] === costHere - 1) {
                const reverseKey = `${nr},${nc},${d}`;
                if (!visitedReverse.has(reverseKey)) {
                    visitedReverse.add(reverseKey);
                    queue.push([nr, nc, d]);
                }
            }
        }

        // Reverse turns
        for (const prevDir of [(d + 3) % 4, (d + 1) % 4]) {
            if (dist[r][c][prevDir] === costHere - 1000) {
                const reverseKey = `${r},${c},${prevDir}`;
                if (!visitedReverse.has(reverseKey)) {
                    visitedReverse.add(reverseKey);
                    queue.push([r, c, prevDir]);
                }
            }
        }
    }

    let tilesOnBestPath = 0;
    for (let r = 0; r < rows; r++) {
        for (let c = 0; c < cols; c++) {
            if (onBestPath[r][c]) tilesOnBestPath++;
        }
    }

    return tilesOnBestPath;
}

/**
 * Main function
 */
function main(): void {
    try {
        const mazeLines = parseInput("\\\\vmware-host\\Shared Folders\\C\\advent-of-code-002\\input-files\\2024\\2024-016\\input.txt");
        const part1Result = solveMaze(mazeLines);
        console.log(`Lowest possible score (Part 1): ${part1Result}`);

        const part2Result = solvePart2(mazeLines);
        console.log(`Number of tiles on at least one best path (Part 2): ${part2Result}`);
    } catch (error) {
        console.error(error);
    }
}

main();

File: Kotlin/2024/2024-016/2024-016.kt
import java.util.PriorityQueue
import java.util.ArrayDeque

data class State(val cost: Int, val row: Int, val col: Int, val direction: Int): Comparable<State> {
    override fun compareTo(other: State): Int {
        return this.cost - other.cost
    }
}

fun solveMaze(mazeLines: List<String>): Int? {
    // Directions: 0=East, 1=South, 2=West, 3=North
    val directions = listOf(
        Pair(0, 1),   // East
        Pair(1, 0),   // South
        Pair(0, -1),  // West
        Pair(-1, 0)   // North
    )

    val rows = mazeLines.size
    val cols = mazeLines[0].length

    var start: Pair<Int, Int>? = null
    var end: Pair<Int, Int>? = null

    for (r in 0 until rows) {
        for (c in 0 until cols) {
            when (mazeLines[r][c]) {
                'S' -> start = Pair(r, c)
                'E' -> end = Pair(r, c)
            }
        }
    }

    if (start == null || end == null) {
        throw IllegalArgumentException("Could not find 'S' or 'E' in the maze.")
    }

    val INF = Int.MAX_VALUE
    val dist = Array(rows) { Array(cols) { IntArray(4) { INF } } }
    val startDir = 0 // facing East
    dist[start.first][start.second][startDir] = 0

    val pq = PriorityQueue<State>()
    pq.add(State(0, start.first, start.second, startDir))

    val visited = mutableSetOf<Triple<Int, Int, Int>>()

    while (pq.isNotEmpty()) {
        val (cost, r, c, d) = pq.poll()

        if (Pair(r, c) == end) {
            return cost
        }

        if (Triple(r, c, d) in visited) continue
        visited.add(Triple(r, c, d))

        val currentDist = dist[r][c][d]
        if (cost > currentDist) continue

        // 1) Move forward (cost + 1) if not a wall
        val (dr, dc) = directions[d]
        val nr = r + dr
        val nc = c + dc
        if (nr in 0 until rows && nc in 0 until cols && mazeLines[nr][nc] != '#') {
            val newCost = cost + 1
            if (newCost < dist[nr][nc][d]) {
                dist[nr][nc][d] = newCost
                pq.add(State(newCost, nr, nc, d))
            }
        }

        // 2) Turn left (cost + 1000)
        val leftDir = (d - 1 + 4) % 4
        val leftCost = cost + 1000
        if (leftCost < dist[r][c][leftDir]) {
            dist[r][c][leftDir] = leftCost
            pq.add(State(leftCost, r, c, leftDir))
        }

        // 3) Turn right (cost + 1000)
        val rightDir = (d + 1) % 4
        val rightCost = cost + 1000
        if (rightCost < dist[r][c][rightDir]) {
            dist[r][c][rightDir] = rightCost
            pq.add(State(rightCost, r, c, rightDir))
        }
    }

    return null // If no path found
}

fun solvePart2(mazeLines: List<String>): Int {
    val rows = mazeLines.size
    val cols = mazeLines[0].length

    // Directions: 0=East, 1=South, 2=West, 3=North
    val directions = listOf(
        Pair(0, 1),   // East
        Pair(1, 0),   // South
        Pair(0, -1),  // West
        Pair(-1, 0)   // North
    )

    var start: Pair<Int, Int>? = null
    var end: Pair<Int, Int>? = null

    for (r in 0 until rows) {
        for (c in 0 until cols) {
            when (mazeLines[r][c]) {
                'S' -> start = Pair(r, c)
                'E' -> end = Pair(r, c)
            }
        }
    }

    if (start == null || end == null) {
        throw IllegalArgumentException("Could not find 'S' or 'E' in the maze.")
    }

    val INF = Int.MAX_VALUE
    val dist = Array(rows) { Array(cols) { IntArray(4) { INF } } }
    val startDir = 0 // facing East
    dist[start.first][start.second][startDir] = 0

    val pq = PriorityQueue<State>()
    pq.add(State(0, start.first, start.second, startDir))
    val visited = mutableSetOf<Triple<Int, Int, Int>>()

    while (pq.isNotEmpty()) {
        val (cost, r, c, d) = pq.poll()

        if (Triple(r, c, d) in visited) continue
        visited.add(Triple(r, c, d))

        val currentDist = dist[r][c][d]
        if (cost > currentDist) continue

        // 1) Move forward
        val (dr, dc) = directions[d]
        val nr = r + dr
        val nc = c + dc
        if (nr in 0 until rows && nc in 0 until cols && mazeLines[nr][nc] != '#') {
            val newCost = cost + 1
            if (newCost < dist[nr][nc][d]) {
                dist[nr][nc][d] = newCost
                pq.add(State(newCost, nr, nc, d))
            }
        }

        // 2) Turn left
        val leftDir = (d - 1 + 4) % 4
        val leftCost = cost + 1000
        if (leftCost < dist[r][c][leftDir]) {
            dist[r][c][leftDir] = leftCost
            pq.add(State(leftCost, r, c, leftDir))
        }

        // 3) Turn right
        val rightDir = (d + 1) % 4
        val rightCost = cost + 1000
        if (rightCost < dist[r][c][rightDir]) {
            dist[r][c][rightDir] = rightCost
            pq.add(State(rightCost, r, c, rightDir))
        }
    }

    val minCostEnd = (0 until 4).minOfOrNull { dist[end.first][end.second][it] } ?: INF
    if (minCostEnd == INF) return 0

    val onBestPath = Array(rows) { BooleanArray(cols) { false } }
    val queue = ArrayDeque<Triple<Int, Int, Int>>()

    for (d in 0 until 4) {
        if (dist[end.first][end.second][d] == minCostEnd) {
            queue.add(Triple(end.first, end.second, d))
        }
    }

    val visitedRev = mutableSetOf<Triple<Int, Int, Int>>()
    visitedRev.addAll(queue)

    while (queue.isNotEmpty()) {
        val (r, c, d) = queue.removeFirst()
        onBestPath[r][c] = true

        val costHere = dist[r][c][d]

        val (dr, dc) = directions[d]
        val rPrev = r - dr
        val cPrev = c - dc
        if (rPrev in 0 until rows && cPrev in 0 until cols && mazeLines[rPrev][cPrev] != '#') {
            if (dist[rPrev][cPrev][d] == costHere - 1 && Triple(rPrev, cPrev, d) !in visitedRev) {
                visitedRev.add(Triple(rPrev, cPrev, d))
                queue.add(Triple(rPrev, cPrev, d))
            }
        }

        for (dPre in listOf((d - 1 + 4) % 4, (d + 1) % 4)) {
            if (dist[r][c][dPre] == costHere - 1000 && Triple(r, c, dPre) !in visitedRev) {
                visitedRev.add(Triple(r, c, dPre))
                queue.add(Triple(r, c, dPre))
            }
        }
    }

    return onBestPath.sumOf { row -> row.count { it } }
}

fun main() {
    val lines = java.io.File("input.txt").readLines()
    val part1Answer = solveMaze(lines)
    if (part1Answer == null) {
        println("No path found!")
    } else {
        println("Lowest possible score: $part1Answer")
    }

    val part2Answer = solvePart2(lines)
    println("Number of tiles on at least one best path: $part2Answer")
}